### DISCUSSION TITLE
How can I run clangd without the presence of code editor (for nightly routine)?

### DISCUSSION THREAD

**Discussion by theshuaiyuan** [clangd/clangd#1667] [General]
I'm using clangd for C++ code projects with VSCode as my code editor. Every once in a while, I need to update my project from repository. Then I generate compile_commands.json file, and use VSCode to open root folder of the project. Next I need some time before clangd can fully parse the entire project.

Since the code is hosted in a remote Linux machine, I'm wondering if I can do this in remote machine without VSCode. For example, I can pull new changes to the project, and compile/generate compile_commands.json file. And instead of using VSCode to open project, can I use clangd to accept compile_commands.json database file, and automatically parse the files in the project, and generate .cache folder? If this is feasible, it will facilitate my development because I don't need to wait my time to finish the parsing?

BTW, the argument I passed from VSCode to clangd is: --background-index --sync -j=32
Thanks

**Comment by HighCommander4** [comment]
Have a look at https://clangd.llvm.org/guides/remote-index

This approach does have the slight difference (and downside) compared to what you describe that it generates the index in a different format than the background indexer, so after the remote indexing is complete, clangd won't incrementally update the resulting index, and you have to periodically regenerate it from scratch. We have https://github.com/clangd/clangd/issues/587 open to track something more like what you describe.

**Comment by theshuaiyuan** [comment]
As of clangd 22.1.0, clangd still lacks the functionality I need. What I mean is, yes, clangd-indexer can generate a single static index file, and my code editor (vscode) can use it for decent go-to-definition and project-wide symbol search. However, find-all-references often returns incomplete results.

clangd-index-server reportedly works much better, but it requires me to manually manage the server's lifecycle rather than having it started and stopped by the editor. And if you have multiple projects that need clangd, you need different ports for different servers, managing multiple server instances — that doesn't feel like what I want either.

So I couldn't help but think of this post from 3 years ago, which seems to have received very few replies. If you stumbled upon or searched your way to this thread, I'd like to share a workaround that has become part of my nightly build: https://github.com/theshuaiyuan/clangd_offline_index

It's a script that mimics an LSP client. It starts a clangd server with `--background-index -j16`, picks one file from `compile_commands.json` to send a `didOpen` notification, which triggers clangd to index all files in the compilation database across the entire project, caching the shards under the `.cache` directory. By the time I open my editor in the morning, the project is already fully indexed and ready to go.

The functionality this workaround provides is quite straightforward. Clangd team, why not just build this into clangd itself?

**Reply by HighCommander4** [reply]
> Clangd team, why not just build this into clangd itself?

Something similar to this is being worked on in https://github.com/llvm/llvm-project/pull/175209.

(Also, just a reminder that there is no "clangd team". There are just volunteers who contribute to the clangd open source project, one of whom could be you.)

**Reply by theshuaiyuan** [reply]
> > Clangd team, why not just build this into clangd itself?
> 
> Something similar to this is being worked on in [llvm/llvm-project#175209](https://github.com/llvm/llvm-project/pull/175209).
> 
> (Also, just a reminder that there is no "clangd team". There are just volunteers who contribute to the clangd open source project, one of whom could be you.)

Thanks for the info, I'll keep an eye on #175209, that's great to know!

I have a lot of respect for everyone who volunteers their time to make clangd as good as it is. It's genuinely one of my favorite tools.
If #175209 lands, I'd be happy to retire my workaround.
